crates/net/src/rlpx/ecies.rs

```diff
@@ -0,0 +1,386 @@
+use aes::cipher::{KeyIvInit, StreamCipher};
+use bytes::BufMut;
+use ethereum_rust_core::{
+    rlp::{
+        decode::RLPDecode,
+        encode::RLPEncode,
+        error::RLPDecodeError,
+        structs::{Decoder, Encoder},
+    },
+    Signature, H128, H512, H520,
+};
+use k256::{
+    ecdsa::SigningKey,
+    elliptic_curve::sec1::{FromEncodedPoint, ToEncodedPoint},
+    EncodedPoint, PublicKey, SecretKey,
+};
+use keccak_hash::H256;
+use rand::{thread_rng, Rng};
+
+type Aes128Ctr64BE = ctr::Ctr64BE<aes::Aes128>;
+
+#[derive(Debug, Clone, Copy)]
+#[allow(unused)]
+pub(crate) struct HandshakeData {
+    pub remote_nonce: H256,
+    pub aes_key: H256,
+    pub mac_key: H256,
+}
+
+#[derive(Debug)]
+// TODO: refactor into two parts, one for the handshake and another one after that.
+pub(crate) struct RLPxConnection {
+    pub nonce: H256,
+    pub ephemeral_key: SecretKey,
+    pub handshake_data: Option<HandshakeData>,
+}
+
+impl RLPxConnection {
+    pub fn new(nonce: H256, ephemeral_key: SecretKey) -> Self {
+        Self {
+            nonce,
+            ephemeral_key,
+            handshake_data: None,
+        }
+    }
+
+    pub fn random() -> Self {
+        let mut rng = thread_rng();
+        Self::new(H256::random_using(&mut rng), SecretKey::random(&mut rng))
+    }
+
+    pub fn is_connected(&self) -> bool {
+        self.handshake_data.is_some()
+    }
+
+    pub fn encode_auth_message(
```

Comment:
Maybe this function could benefit from some inline comments

Reply:
Added!

Reply:
Awesome!